ibm/data_source_ibm_dl_gateway.go

```diff
@@ -0,0 +1,475 @@
+
+		}
+		if instance.Status != nil {
+			gatewayVC[dlVCStatus] = *instance.Status
+		}
+
+		gatewayVCs = append(gatewayVCs, gatewayVC)
+	}
+	d.SetId(dlGatewayId)
+
+	d.Set(dlGatewaysVirtualConnections, gatewayVCs)
+	return nil
+}
+func dataSourceIBMDLGatewayRead(d *schema.ResourceData, meta interface{}) error {
+	directLink, err := directlinkClient(meta)
+	dlGatewayName := d.Get(dlName).(string)
+
+	if err != nil {
+		return err
+	}
+	listGatewaysOptionsModel := &directlinkapisv1.ListGatewaysOptions{}
+	listGateways, _, err := directLink.ListGateways(listGatewaysOptionsModel)
+	if err != nil {
+		return err
+	}
+	var dlGatewayId string
+
+	for _, instance := range listGateways.Gateways {
+
+		if *instance.Name == dlGatewayName {
+
+			if instance.ID != nil {
+				d.SetId(*instance.ID)
+				dlGatewayId = *instance.ID
+			}
+			if instance.Name != nil {
+				d.Set(dlName, *instance.Name)
+			}
+			if instance.Crn != nil {
+				d.Set(dlCrn, *instance.Crn)
+			}
+			if instance.BgpAsn != nil {
+				d.Set(dlBgpAsn, *instance.BgpAsn)
+			}
+			if instance.BgpIbmCidr != nil {
+				d.Set(dlBgpIbmCidr, *instance.BgpIbmCidr)
+			}
+			if instance.BgpIbmAsn != nil {
+				d.Set(dlBgpIbmAsn, *instance.BgpIbmAsn)
+			}
+			if instance.Metered != nil {
+				d.Set(dlMetered, *instance.Metered)
+			}
+			if instance.CrossConnectRouter != nil {
+				d.Set(dlCrossConnectRouter, *instance.CrossConnectRouter)
+			}
+			if instance.BgpBaseCidr != nil {
+				d.Set(dlBgpBaseCidr, *instance.BgpBaseCidr)
+			}
+			if instance.BgpCerCidr != nil {
+				d.Set(dlBgpCerCidr, *instance.BgpCerCidr)
+			}
+			if instance.DedicatedHostingID != nil {
+				d.Set(dlDedicatedHostingID, instance.DedicatedHostingID)
+			}
+			if instance.ProviderApiManaged != nil {
+				d.Set(dlProviderAPIManaged, *instance.ProviderApiManaged)
+			}
+			if instance.Type != nil {
+				d.Set(dlType, *instance.Type)
+			}
+			if instance.SpeedMbps != nil {
+				d.Set(dlSpeedMbps, *instance.SpeedMbps)
+			}
+			if instance.OperationalStatus != nil {
+				d.Set(dlOperationalStatus, *instance.OperationalStatus)
+			}
+			if instance.BgpStatus != nil {
+				d.Set(dlBgpStatus, *instance.BgpStatus)
+			}
+			if instance.LocationName != nil {
+				d.Set(dlLocationName, *instance.LocationName)
+			}
+			if instance.LocationDisplayName != nil {
+				d.Set(dlLocationDisplayName, *instance.LocationDisplayName)
+			}
+			if instance.Vlan != nil {
+				d.Set(dlVlan, *instance.Vlan)
+			}
+			if instance.Global != nil {
+				d.Set(dlGlobal, *instance.Global)
+			}
+			if instance.Port != nil {
+				d.Set(dlPort, *instance.Port.ID)
+			}
+			if instance.LinkStatus != nil {
+				d.Set(dlLinkStatus, *instance.LinkStatus)
+			}
+			if instance.CreatedAt != nil {
+				d.Set(dlCreatedAt, instance.CreatedAt.String())
+			}
+			if instance.ResourceGroup != nil {
+				rg := instance.ResourceGroup
+				d.Set(dlResourceGroup, *rg.ID)
+			}
+
+		}
+	}
+
+	if dlGatewayId == "" {
```

Comment:
instead dlGatewayId declare a local bool variable found (false) 
move this inside the if block match
```
	d.SetId(dlGatewayId)
	return dataSourceIBMDLGatewayVirtualConnectionsRead(d, meta)
```
and outside for loop add a if !found {
		return fmt.Errorf(
			"Error Gateway with name  (%s) not found ", dlGatewayName)
}